File: 662. Maximum Width of Binary Tree/Solution.py
# https://leetcode.com/problems/maximum-width-of-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
	def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
		# Create a deque called level, and add root to it with its index as 1
		level = deque([(root, 1)])
		
		# Initialize res to 0
		res = 0
		
		# While level is not empty
		while(level):
			# Create a new deque called newlevel
			newlevel = deque()

			# Update res to the maximum value of res and the difference between the index of the last and first nodes in level, plus one (which is the width of the current level)
			res = max(res, level[-1][1] - level[0][1] + 1)

			# Iterate through the current level
			while level:
				# Remove the leftmost node and its index from level
				current_node, ind = level.popleft()

				# If the left child of the current node exists, add it to newlevel with an index of twice the index of the current node
				if(current_node.left):
					newlevel.append((current_node.left, 2*ind))
				
				# If the right child of the current node exists, add it to newlevel with an index of twice the index of the current node plus one
				if(current_node.right):
					newlevel.append((current_node.right, 2*ind + 1))

			# Replace level with newlevel
			level = newlevel

		# Return the result
		return res

File: 662. Maximum Width of Binary Tree/Solution.cpp
// https://leetcode.com/problems/maximum-width-of-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	/*
		Time complexity: O(N), where N is the number of nodes in the binary tree, because each node is visited once.

		Space complexity: O(M), where M is the maximum number of nodes that can be present at a single level in the binary tree. This is because, at most, all the nodes at the maximum level can be stored in the "level" deque at the same time.
	*/
    int widthOfBinaryTree(TreeNode* root) {
        // Create a deque to store nodes at each level along with their corresponding index
        deque<tuple<TreeNode*,long long>> level;
        level.push_back({root, 1});

        // Initialize the result variable to store the maximum width of the binary tree
        int res = 0;

        // Continue processing levels until there are no more nodes
        while (!level.empty()){
            // Create a new deque to store nodes at the next level
            deque<tuple<TreeNode*,long long>> newlevel;

            // Calculate the width of the current level by subtracting the index of the first node
            // from the index of the last node, and update the result if necessary
            if(get<1>(level.back()) - get<1>(level.front()) + 1 > res){
                res = get<1>(level.back()) - get<1>(level.front()) + 1;
            }

            // Process nodes at the current level
            while (!level.empty()){
                // Get the current node and its index
                TreeNode* current_node = get<0>(level.front());
                int ind = get<1>(level.front());
                level.pop_front();

                // Add the left child of the current node to the newlevel deque, along with its updated index
                if(current_node->left != NULL){
                    newlevel.push_back({current_node->left, (long long) 2*ind});
                }

                // Add the right child of the current node to the newlevel deque, along with its updated index
                if(current_node->right != NULL){
                    newlevel.push_back({current_node->right, (long long) 2*ind + 1});
                }
            }

            // Update the level deque with the nodes at the next level
            level = newlevel;
        }

        // Return the maximum width of the binary tree
        return res;
    }
};
